common/workitem.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright © 2017 Drew Walters <drewwalters96@gmail.com>
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <debug.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <workitem.h>
+
+/* Work item queue capacity. */
+#define MAX_WORK_ITEMS 1
+
+/* Queue of work items. */
+static struct work_item work_items[MAX_WORK_ITEMS];
+
+void
+queue_work_item(work_function work_fn, void *work_fn_params)
+{
+	struct work_item *queued;
+	struct work_item  work;
+
+	queued = NULL;
+
+	/* Find first available index and verify work item is not queued. */
+	for (size_t i = 0; i < MAX_WORK_ITEMS; i++) {
+		if (!work_items[i].work_fn && !queued) {
+			work.work_fn  = work_fn;
+			work.params   = work_fn_params;
+			work_items[i] = work;
+			queued        = &work_items[i];
+		} else if (work_items[i].work_fn == work_fn &&
```

Comment:
There could be potentially be times where there are vaild duplicate work items in the queue. I beleive this should be removed.

Reply:
The total number of possible work items must be bounded. This isn't possible if we allow duplicates. The current code still allows where the same function needs to be run with different parameters (e.g. for different devices).